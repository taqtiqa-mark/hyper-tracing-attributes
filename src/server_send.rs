pub use analyze::{analyze, Model};
pub use codegen::codegen;
pub use lower::{lower, Ir};
pub use parse::{parse, parse_args, Ast};

pub mod analyze;
pub mod codegen;
pub mod lower;
pub mod parse;
