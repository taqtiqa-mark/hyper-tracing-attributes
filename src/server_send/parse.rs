#![allow(dead_code)]

use proc_macro2::TokenStream;
use proc_macro_error::{abort, abort_call_site};
use syn::{Expr, Item};

pub type Ast = syn::ItemFn;

// Standard parsing of input arguments that are assignment expressions.
#[derive(Debug)]
pub struct Args {
    pub vars: Vec<syn::Expr>
}

// Parse the surrounding attribute data once we know it contains ExprAssign
impl syn::parse::Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
        // parses a=1,b="d",c=f(), into Expr
        let vars = syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated(input)?;
        Ok(Args {
            vars: vars.into_iter().collect(),
        })
    }
}

// Parse attribute arguments to assess correctness.
// This utility function generates a no-op/dummy output, because it is
// required by the fact `syn::parse_macro_input!` must be called in a
// function that returns `proc_macro::TokenStream`.
pub fn parse_args(metadata: proc_macro::TokenStream) -> proc_macro::TokenStream {
    const ERROR: &str = "this attribute takes from one to three assign arguments";
    const HELP: &str = "use `#[server_send(level=Level::WARN, name=\"Name\", skip=[a, b])]`";

    if metadata.is_empty() {
        // ../tests/ui/server_send/error/has-no-arguments.rs
        abort_call_site!(ERROR; help = HELP)
    }

    let md = metadata.clone();
    let args = syn::parse_macro_input!(md as Args);

    if args.vars.len() > 3 {
        // ../tests/ui/server_send/error/has-too-many-arguments.rs
        abort_call_site!(ERROR; help = HELP)
    }

    for expr in &args.vars {
        match expr {
            Expr::Array(expr) => {
                // ../tests/ui/server_send/error/has-expression-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
            Expr::Assign(_expr) => { }
            _ => {
                // ../tests/ui/server_send/error/has-other-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
        }
    }
    proc_macro::TokenStream::from(quote::quote! {fn dummy(){}})
}

pub fn parse(metadata: TokenStream, item: TokenStream) -> Ast {

    parse_args(metadata.into());

    match syn::parse2::<Item>(item) {
        Ok(Item::Fn(item)) => item,
        Ok(item) => {
            // ../tests/ui/item-is-not-a-function.rs
            abort!(
                item,
                "item is not a function";
                help = "`#[server_send(...)]` can only be used on functions"
            )
        }
        Err(_) => unreachable!(), // ?
    }
}
