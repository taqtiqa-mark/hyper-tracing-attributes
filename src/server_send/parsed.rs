#![allow(dead_code)]

use proc_macro2::TokenStream;
use proc_macro_error::{abort, abort_call_site};
use syn::{Expr, Item, ItemFn};
use darling;

pub type Ast = ItemFn;

/// Describes the data that directly corresponds to the attributes of the
/// `server_send`-proc-macro.
/// This is the parsing target for the
/// crate "darling" which works together with `syn`.
/// FromMeta implements "from_list"-factory-method (`darling` crate).
#[derive(Debug, darling::FromMeta)]
struct TaskAttributes {
    #[darling(default)]
    level: Option<String>,
    #[darling(default)]
    name: Option<String>,
    #[darling(default)]
    skip: Option<String>,
    #[darling(default)]
    fields: Option<String>,
}

// #[derive(Debug, darling::FromMeta)]
// struct InstrumentAttributes {
//     #[darling(default)]
//     level: Option<syn::Expr>,
//     #[darling(default)]
//     name: Option<syn::Expr>,
//     #[darling(default)]
//     skip: Option<syn::Expr>,
//     #[darling(default)]
//     fields: Option<syn::Expr>,
// }

impl TaskAttributes {
    /// Maps property `skip` of `TaskAttributes` from a
    /// comma-separated string to a vector of strings.
    fn skip_params(&self) -> Vec<String> {
        self.skip
            .as_ref()
            .map(|s| s.clone())
            .map(|s| {
                s.split(",")
                    .into_iter()
                    .map(|s| s.to_owned())
                    .collect::<Vec<String>>()
            })
            .unwrap_or(vec![])
            .into_iter()
            .map(|s| s.trim().to_owned())
            .collect::<Vec<String>>()
    }

    /// Maps property `read` of `TaskAttributes` from a
    /// comma-separated string to a vector of strings.
    fn fields_params(&self) -> Vec<String> {
        self.fields
            .as_ref()
            .map(|s| s.clone())
            .map(|s| {
                s.split(",")
                    .into_iter()
                    .map(|s| s.to_owned())
                    .collect::<Vec<String>>()
            })
            .unwrap_or(vec![])
            .into_iter()
            .map(|s| s.trim().to_owned())
            .collect::<Vec<String>>()
    }
}

// /// Invoked like this:
// /// - `#[task(write = "data1, data2")]`
// /// - `#[task(write = "data1", read = "data2")]`
// /// - `#[task(read = "data2")]`
// /// If a parameter is write it is automatically also read.
// #[proc_macro_attribute]
// pub fn task(args: TokenStream, item: TokenStream) -> TokenStream {
//     // we do nothing here; just a marker for the super macro
//     let attr_args = parse_macro_input!(args as AttributeArgs);
//     let attrs: TaskAttributes =
//         TaskAttributes::from_list(&attr_args).expect("#[task] could not be parsed!");
//     println!("read_params: {:#?}", attrs.read_params());
//     println!("write_params: {:#?}", attrs.write_params());
//     item
// }

// Attribute arguments that are assignment expressions.
#[derive(Debug)]
pub struct Args {
    vars: Vec<syn::Expr>
}

// Parse the attribute data once we know it contains ExprAssign
impl syn::parse::Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
        // parses a=1,b="d",c=f(), where a,b and c are ExprAssign
        let vars = syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated(input)?;
        Ok(Args {
            vars: vars.into_iter().collect(),
        })
    }
}

// Parse attribute arguments then count.
// This utility/no-op function is required by the fact
// `syn::parse_macro_input!` must be called in a function that returns
// proc_macro::TokenStream.
pub fn parse_args(metadata: proc_macro::TokenStream) -> proc_macro::TokenStream {
    const ERROR: &str = "this attribute takes from one to four arguments";
    const HELP: &str = "use `#[server_send(level=trace, name=\"Name\", skip=\"a, b\")]`";

    // eprintln!("{:?}", metadata);

    if metadata.is_empty() {
        // ../tests/ui/server_send/error/has-no-arguments.rs
        abort_call_site!(ERROR; help = HELP)
    }

    let md = metadata.clone();
    let args = syn::parse_macro_input!(md as Args);

    // eprintln!("{}",args.vars.len());

    // eprintln!("{:?}", args);

    for expr in &args.vars {
        match expr {
            Expr::Array(expr) => {
                // ../tests/ui/server_send/error/has-expression-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
            Expr::Assign(_expr) => { }
            _ => {
                // ../tests/ui/server_send/error/has-other-arguments.rs
                abort!(expr, ERROR; help = HELP)
            }
        }
    }
    proc_macro::TokenStream::from(quote::quote! {fn dummy(){}})
}

pub fn parse(metadata: TokenStream, item: TokenStream) -> Ast {

    parse_args(metadata.clone().into());

    match syn::parse2::<Item>(item) {
        Ok(Item::Fn(item)) => item,
        Ok(item) => {
            // ../tests/ui/server_send/error/item-is-not-a-function.rs
            abort!(
                item,
                "item is not a function";
                help = "`#[server_send(...)]` can only be used on functions"
            )
        }
        Err(_) => unreachable!(), // ?
    }
}

// #[cfg(test)]
// mod tests {
//     use quote::quote;

//     use super::*;

//     #[test]
//     fn valid_syntax() {
//         parse(
//             quote!(),
//             quote!(
//                 #[inline]
//                 #[server_send(level = debug)]
//                 fn even_to_odd(x: u32) -> u32 {
//                     x + 1
//                 }
//             ),
//         );
//     }
// }
