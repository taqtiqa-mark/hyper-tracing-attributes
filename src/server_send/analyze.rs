use proc_macro_error::{abort};
use syn::{
    parenthesized,
    parse::{Parse, ParseStream},
    spanned::Spanned,
    ItemFn,
};

use crate::server_send::Ast;

pub fn analyze(ast: Ast) -> Model {
    let mut fields = vec![];
    let mut level: syn::ExprPath = syn::parse_quote!(Level::DEBUG);
    let mut skip: syn::ExprArray = syn::parse_quote!([]);

    let mut item = ast;
    let attrs = &mut item.attrs;
    for index in (0..attrs.len()).rev() {
        eprintln!("{:#?}", attrs[index]);
        if let Some(ident) = attrs[index].path.get_ident() {
            let id = ident.to_string();
            match id.as_str() {
                "server_send" | "trace_field" => {
                    let attr = attrs.remove(index);
                    let span = attr.tokens.span();
                    if id == "server_send" {
                        get_level(&mut level, span, attr.tokens.clone());
                        get_skip(&mut skip, span, attr.tokens.clone());
                    }
                    if id == "trace_field" {
                        get_field(&mut fields, span, attr.tokens);
                    }
                }
                _ => {}
            }
        }
    }

    Model {
        fields,
        item,
        level,
        skip,
    }
}

fn get_field(fields: &mut Vec<syn::ExprAssign>, span: proc_macro2::Span, tokens: proc_macro2::TokenStream) {
    if let Ok(arg) = syn::parse2::<AttributeArgument>(tokens) {
        fields.push(arg.expr);
    } else {
        // ../tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs
        abort!(
            span,
            "expected an assigned expression as argument";
            help = "example: `#[trace_field(argument = 0)]`")
    }
}

fn get_level(level: &mut syn::ExprPath, span: proc_macro2::Span, tokens: proc_macro2::TokenStream) {

    if let Ok(arg) = syn::parse2::<AttributeArgument>(tokens) {
        if let syn::Expr::Path(path) = *(arg.expr).right {
            *level = path;
        }
    } else {
        // ../tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs
        abort!(
            span,
            "expected an assigned expression as argument";
            help = "example: `#[server_send(level = tracing::Level::INFO)]`")
    }
}

fn get_skip(skip: &mut syn::ExprArray, span: proc_macro2::Span, tokens: proc_macro2::TokenStream) {

    if let Ok(arg) = syn::parse2::<AttributeArgument>(tokens) {
        if let syn::Expr::Array(ary) = *(arg.expr).right {
            *skip = ary;
        }
    } else {
        // ../tests/ui/server_send/error/trace_field-is-not-an-assignment-expression.rs
        abort!(
            span,
            "expected an assigned expression as argument";
            help = "example: `#[server_send(skip = [a,b,c])]`")
    }
}

struct AttributeArgument {
    expr: syn::ExprAssign,
}

impl Parse for AttributeArgument {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let content;
        let _parenthesis = parenthesized!(content in input);

        Ok(AttributeArgument {
            expr: content.parse()?,
        })
    }
}

#[derive(Debug)]
pub struct Model {
    pub fields: Vec<syn::ExprAssign>,
    pub item: ItemFn,
    pub level: syn::ExprPath,
    pub skip: syn::ExprArray,
}

#[cfg(test)]
mod tests {
    use syn::{parse_quote, Attribute};

    use super::*;

    #[test]
    fn can_extract_trace_field() {
        let model = analyze(parse_quote!(
            #[trace_field(x=0)]
            fn f(x: bool) {}
        ));

        let expected: &[syn::ExprAssign] = &[parse_quote!(x=0)];
        assert_eq!(expected, model.fields);

        assert!(model.item.attrs.is_empty());
    }

    #[test]
    fn can_extract_level() {
        let model = analyze(parse_quote!(
            #[server_send(level=Level::DEBUG)]
            fn f(x: bool) {}
        ));

        let expected: syn::ExprPath = parse_quote!(Level::DEBUG);
        assert_eq!(expected, model.level);

        assert!(model.item.attrs.is_empty());
    }

    #[test]
    fn can_extract_skip() {
        let model = analyze(parse_quote!(
            #[server_send(skip=[x])]
            fn f(x: bool) {}
        ));

        let expected: syn::ExprArray = parse_quote!([x]);
        assert_eq!(expected, model.skip);

        assert!(model.item.attrs.is_empty());
    }

    // Ensure attributes unrelated to this proc-macro are not removed or reordered
    #[test]
    fn non_dsl_attributes_are_preserved() {
        let model = analyze(parse_quote!(
            #[a]
            #[trace_field(x=0)]
            #[b]
            fn f(x: bool) {}
        ));

        let expected: &[Attribute] = &[parse_quote!(#[a]), parse_quote!(#[b])];
        assert_eq!(expected, model.item.attrs);
    }
}
