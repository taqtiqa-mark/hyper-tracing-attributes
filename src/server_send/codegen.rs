#![allow(unused_variables)]

use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::{parse_quote, ItemFn, Stmt};

use crate::server_send::{lower::Field, Ir};

pub type Rust = TokenStream;

pub fn codegen(ir: Ir) -> Rust {

    let Ir { fields, item, level, name, skip } = ir;

    let ItemFn {
        attrs,
        vis,
        sig,
        block,
    } = item;

    let flds: Vec<syn::ExprAssign> = fields.into_iter().map(|p| p.expr).collect();
    let stmts = &block.stmts;
    // let skp: Vec<syn::Lit> = skip.elems.into_iter().map(|s| {
    //     let syn::ExprLit { attrs, lit} = s;
    //     lit }).collect();
    let skp: Vec<syn::Expr> = skip.elems.into_iter().map(|s| s).collect();

    let code = syn::parse_quote! {
        #(#attrs)*
        #[cfg_attr(feature = "tracing",
            tracing::instrument(level = #level,
                                name = #name,
                                skip(
                                    #(#skp),*
                                ),
                                fields( //Custom
                                        #(#flds ,)*
                                        //
                                        otel.name           = "Server::parse",

                                )
                        )
                    )]
        #vis #sig {
            #(#stmts)*
        }
    };
    eprintln!("{}", code);
    code
}

impl ToTokens for Field {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Field { expr } = self;
        // The trace field line: x = 0,
        let stmt: Stmt = parse_quote!(#expr);
        stmt.to_tokens(tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn output_is_function_item() {
        let pq: syn::ExprAssign = parse_quote!(x=0);
        let i = syn::parse2(quote::quote!(
                fn f() {}
            ).into()).expect("ItemFn");
        let f = vec![Field {
                expr: pq
            }];
        let l = syn::parse_quote!(Level::TRACE);
        let n = syn::parse_quote!("some_test");
        let s = syn::parse_quote!([a,b,c]);
        let ir = Ir {
                fields: f,
                item: i,
                level: l,
                name: n,
                skip: s
            };

        let rust = codegen(ir);

        assert!(syn::parse2::<ItemFn>(rust).is_ok());
    }
}
