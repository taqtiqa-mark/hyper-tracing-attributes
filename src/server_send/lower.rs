use syn::{ItemFn};

use crate::server_send::Model;

pub fn lower(model: Model) -> Ir {
    let Model {
        level,
        fields,
        item,
    } = model;

    let fields = fields
        .into_iter()
        .map(|expr| Field {
            expr,
        })
        .collect();

    let level = level;

    Ir { level, fields, item }
}

#[derive(Debug)]
pub struct Ir {
    pub level: syn::ExprPath,
    pub fields: Vec<Field>,
    pub item: ItemFn,
}

#[derive(Debug)]
pub struct Field {
    pub expr: syn::ExprAssign,
    //pub message: String,
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::*;

    impl Model {
        fn stub() -> Self {
            Self {
                level: parse_quote!(Level::TRACE),
                fields: vec![],
                item: parse_quote!(
                    fn f() {}
                ),
            }
        }
    }

    #[test]
    fn produces_expression_for_field() {
        let mut model = Model::stub();
        model.fields.push(parse_quote!(x=0));

        let ir = lower(model);

        assert_eq!(1, ir.fields.len());

        let field = &ir.fields[0];
        let expected: syn::ExprAssign = parse_quote!(x=0);
        assert_eq!(expected, field.expr);
    }

    #[test]
    fn produces_expression_for_level() {
        // Setup Level::TRACE
        let mut model = Model::stub();
        model.level = parse_quote!(Level::DEBUG);

        let ir = lower(model);

        let level = ir.level;
        let expected: syn::ExprPath = parse_quote!(Level::DEBUG);
        assert_eq!(expected, level);
    }
}
