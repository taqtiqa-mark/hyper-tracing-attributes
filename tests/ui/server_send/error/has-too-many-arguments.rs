use tracing_attributes_http::server_send;

#[allow(unused_braces)]
#[server_send(level=trace, name="B", skip="self", target="bulls-eye")]
fn f(a: u32) -> u32 {
    a
}

fn main() {}
