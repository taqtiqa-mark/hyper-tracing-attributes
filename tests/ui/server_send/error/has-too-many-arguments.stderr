warning: unused variable: `meta_args`
  --> src/lib.rs
   |
   |     let mut meta_args = syn::parse_macro_input!(meta as crate::server_send::parse::Args);
   |             ^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_meta_args`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: variable does not need to be mutable
  --> src/lib.rs
   |
   |     let mut meta_args = syn::parse_macro_input!(meta as crate::server_send::parse::Args);
   |         ----^^^^^^^^^
   |         |
   |         help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

error: this attribute takes from one to three assign arguments

         = help: use `#[server_send(level=Level::WARN, name="Name", skip=[a, b])]`

 --> tests/ui/server_send/error/has-too-many-arguments.rs:4:1
  |
4 | #[server_send(level=trace, name="B", skip="self", target="bulls-eye")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `server_send` (in Nightly builds, run with -Z macro-backtrace for more info)
